### ISSUE TITLE
TCK: Add a TCK test that verifies users can enlist for AfterLRA notification even after the LRA enters ending phase (Closing/Cancelling)

### ISSUE THREAD

**Issue by xstefank** [eclipse/microprofile-lra#241]

**Comment by mmusgrov** [comment]
+1

**Comment by rdebusscher** [comment]
@xstefank Can you indicate how this can work? Also there seems no mention of this in the Spec document.

**Comment by xstefank** [comment]
@mmusgrov mentioned in different issue/PR that this is an implicit requirement. My initial thinking is writing a resource that will be returning 202 ACCEPTED from Compensate/Complete or Status depending on a flag so the flow of the test would be:

* start and close/compensate LRA with resource enlist
* complete/compensate return 202 so the LRA is in the state Closing/Cancelling
* enlist different resource for AfterLRA notification with this ending LRA
* switch the flag in the first resource so on the next invocation Compensate/Complete returns 200 Compensated/Completed
* verify that AfterLRA was called on the second resource

@rdebusscher would that work?

**Comment by rdebusscher** [comment]
@xstefank 
> enlist different resource for AfterLRA notification with this ending LRA

But that different Resource is not a participant then, but just JAX-RS resource with @AfterLRA method.  (Participant need LRAStatus.active)

But a clarification within spec is probably useful.  Now we mention only for @LRA.Type 'when called inside of LRA' and 'when called outside of LRA' and thus we need to define what 'inside' and 'outside' is (with LRA statuses) 

inside = LRAStatus.active, LRAStatus.closing, LRAStatus.cancelling  (and not only active as I interpreted it until now)
outside = other LRAStatus

**Comment by xstefank** [comment]
@rdebusscher the second resource would not be LRA participant as defined in the specification, only AfterLRA observer (still it needs to have LRA method so probably your later comments are still valid. So just for completeness it will have LRA and AfterLRA methods but NO Compensate method).

I will just add that inside will probably also include NESTED LRA in Closed state while its parent LRA is still active as it still can receive AfterLRA notification again when the parent is cancelled?

And I would suggest different naming as we are mixing two different things here. For me, we should distinguish phases where it is possible to still join the LRA (LRA context must be active) and the phases where it is possible to enlist for AfterLRA notifications. I agree that some spec text clarification will be useful.

**Comment by mmusgrov** [comment]
You just need to register an AfterLRA listener because that kind of resource does not need to be a participant (that is already explicitly stated in the spec). But it is an implicit requirement to support this notification at any point up until the LRA is closed or cancelled (but I have suggested that we add a sentence to state that it is possible to receive such notifications).

**Comment by mmusgrov** [comment]
> ... but I have suggested that we add a sentence to state that it is possible to receive such notifications).

But let's not make it too wordy since that will make things look more complicated than it needs to be.
